File: problem-solving/adventofcode2020/day6.py
# https://adventofcode.com/2020/day/6
input_file = open('./day6.txt', 'r')
input_list = input_file.read().splitlines()

def solve_first():
    total_count = 0
    count_set = set()
    for i in input_list:
        if i == "":
            total_count += get_unique_question_count(count_set)
            count_set = set()
            continue
        update_count_set(i, count_set)
    total_count += get_unique_question_count(count_set)
    return total_count

def update_count_set(questions, count_set):
    count_set.update(set(questions))

def get_unique_question_count(count_set):
    return len(count_set)

def solve_second():
    total_count = 0
    count_dict = {}
    group_member_count = 0
    for i in input_list:
        if i == "":
            total_count += sum(value == group_member_count for value in count_dict.values())
            group_member_count = 0
            count_dict = {}
            continue
        group_member_count += 1
        update_count_dict(i, count_dict)
    total_count += sum(value == group_member_count for value in count_dict.values())
    return total_count

def update_count_dict(questions, count_dict):
    for q in list(questions):
        if q in count_dict:
            count_dict[q] += 1
        else:
            count_dict[q] = 1

if __name__ == "__main__":
    print(solve_first())
    print(solve_second())

File: problem-solving/adventofcode2020/day6.cpp
// https://adventofcode.com/2020/day/6
#include <algorithm>
#include <fstream>
#include <iostream>
#include <set>
#include <string>
#include <unordered_map>
#include <vector>
#include <iterator>

using namespace std;

vector<string> readInput(string filepath)
{
    vector<string> inputList;
    ifstream inputFile;
    inputFile.open(filepath);

    string line;
    while (getline(inputFile, line))
    {
        inputList.push_back(line);
    }

    inputFile.close();
    return inputList;
}

int getUniqueQuestionCount(set<char> &countSet) { return countSet.size(); }

void updateCountSet(string input, set<char> &countSet)
{
    for (int i = 0; i < input.length(); i++)
    {
        countSet.insert(input[i]);
    }
}

int solvePart1(const vector<string> &inputList)
{
    int totalCount = 0;
    set<char> countSet;
    for (int i = 0; i < inputList.size(); i++)
    {
        if (inputList[i] == "")
        {
            totalCount += getUniqueQuestionCount(countSet);
            countSet.clear();
            continue;
        }
        updateCountSet(inputList[i], countSet);
    }
    totalCount += getUniqueQuestionCount(countSet);
    return totalCount;
}

void updateCountMap(string input, unordered_map<char, int> &countMap)
{
    for (int i = 0; i < input.length(); i++)
    {
        if (countMap.count(input[i]))
        {
            countMap[input[i]]++;
        }
        else
        {
            countMap.insert({input[i], 1});
        }
    }
}

int solvePart2(const vector<string> &inputList)
{
    int totalCount = 0;
    unordered_map<char, int> countMap;
    int groupMemberCount = 0;
    for (int i = 0; i < inputList.size(); i++)
    {
        string input = inputList[i];
        if (input == "")
        {
            for (unordered_map<char, int>::iterator it = countMap.begin(); it != countMap.end(); ++it)
            {
                totalCount += (int)(it->second == groupMemberCount);
            }
            groupMemberCount = 0;
            countMap.clear();
            continue;
        }
        groupMemberCount++;
        updateCountMap(input, countMap);
    }
    for (unordered_map<char, int>::iterator it = countMap.begin(); it != countMap.end(); ++it)
    {
        totalCount += (int)(it->second == groupMemberCount);
    }
    return totalCount;
}

int main()
{
    vector<string> inputList = readInput("./day6.txt");
    cout << solvePart1(inputList) << endl;
    cout << solvePart2(inputList) << endl;
}
